audience restriction, end2end example, claims parameter, and token processing (#26)
oauthstuff/draft-oauth-rar: main.md

For example, a request for payment authorization can be represented using a JSON object like this:

```JSON
{ 
   "type":"payment_initiation",
   "instructedAmount":{ 
      "currency":"EUR",
      "amount":"123.50"
   },
   "creditorName":"Merchant123",
   "creditorAccount":{ 
      "iban":"DE02100100109307118603"
   },
   "remittanceInformationUnstructured":"Ref Number Merchant"
}
```

This object contains detailed information about the intended payment, such as amount, currency, and creditor, that are required to inform the user and obtain her consent. The AS and the respective RS (providing the payment initation API) will together enforce this consent.

For a comprehensive discussion of the challenges arising from new use cases in the open banking and electronic signing spaces see [@transaction-authorization]. 

In addition to facilitating custom authorization requests, this draft also introduces a set of common data type fields for use across different APIs. 

> Most notably, the field `locations` allows a client to specify where it intends to use a certain authorization. i.e. it is now possible to unambiguously assign permissions to resource servers. In situations with multiple resource servers, this prevents unintended client authorizations (e.g. a `read` scope value potentially applicable for an email as well as a cloud service). In combination with the `resource` token request parameter as specified in [@I-D.ietf-oauth-resource-indicators] it enables the AS to mint RS-specfic structured access tokens that only contain the permissions applicable to the respective RS.

Author: added locations element to example

"grant type", "access token request", "access token response", and
"client" defined by The OAuth 2.0 Authorization Framework [@!RFC6749].

# Request parameter "authorization_details" {#authz_details}

> The request parameter `authorization_details` contains, in JSON notation, an array of objects. Each JSON object contains the data to specify the authorization requirements for a certain type of resource. The type of resource or access requirement is determined by the `type` field. 

Reviewer: ```suggestion
The request parameter `authorization_details` is a JSON array that contains JSON objects. Each JSON object contains the data to specify the authorization requirements for a certain type of resource. The type of resource or access requirement is determined by the `type` field. 
```

Reviewer: I like the explicit phrasing of the original text, but I think this reads a little better.

The following example shows how an implementation could utilize the namespace `https://scheme.example.org/` to ensure collision resistant element names.

```JSON
{
   "type": "https://scheme.example.org/files",
   "locations": [
      "https://example.com/files"
   ],
   "permissions": [
      {
         "path": "/myfiles/A",
         "access": [
            "read"
         ]
      },
      {
         "path": "/myfiles/A/X",
         "access": [
            "read",
            "write"
         ]
      }
   ]
}
```
## Relationship to "scope" parameter {#scope}

`authorization_details` and `scope` can be used in the same authorization request for carrying independent authorization requirements. 

> The AS MUST keep both requirements sets separate when reflecting this data to the client and passing this data through to resource servers.

Reviewer: Why is this a requirement? It might make sense for the AS to treat a `scope` value as a shorthand for one of the `authorization_details` objects, or even a set of them. I think it's more important that the AS always merge them together to create the whole request. In other words, it's an AND.

Author: Do you intend to define a data structure for conveying the merged result?

I tend towards keeping the way scopes are passed through in JWTs and Introspection Responses as is and define an additional pipe for authorization details.

Reviewer: It was not my intent to define a merged data structure, but having guidance for how to combine these two is going to be one of the harder parts of merging this in with OAuth 2's legacy code.

Author: Can you please propose text?

Reviewer: ```suggestion
The AS MUST consider both sets of requirements in combination with each other for the given authorization request. The details of how the AS combines these parameters are specific to the APIs being protected and outside the scope of this specification.

It is RECOMMENDED that a given API use only one form of requirement specification. 
```

> For user convenience, the AS is supposed to merge the requirements when asking the user for consent.

Reviewer: This isn't just for convenience, this is for security: if I ask for a `scope` that's got more access than the `details` object asks for, it should be that the result is both of them together. Otherwise we can end up with naive implementations giving more than they intend, I think.

Author: Even if the AS does not merge the authz requirements it must ask for consent for every detail. I don't see how this could result in security issues.

Besides this, the text states that scopes and authz details shall be independent. So you can use openid as scope but the authz details shall not contain openid stuff. I know that's pretty restrictive but I also think it's acceptable and (I hope) makes things easier.

Reviewer: I don't think that's  a testable restriction, to be honest. And we're going to end up with some API that wants to break that. 

The security issue comes from potentially opening things wider than one intended.

Author: So your proposal is we should accept that implementers use both channels to pass requirements re the same API.  I'm fine with it as long as we can define what this means. Can you please propose text?

Reviewer: ```suggestion
When gathering user consent, the AS MUST present the merged set of requirements represented by the authorization request. 
```

Reviewer 2: How about "complete" rather than "merged"? Trying to say that you gotta show em all somehow but not necessarily saying how. The word "merged" to me kinda suggests some processing.

Reviewer: ```suggestion
When gathering user consent, the AS MUST present the complete set of requirements represented by the authorization request. 
```

Reviewer: I also like "complete" better:

```suggestion
When gathering user consent, the AS MUST present the complete set of requirements represented by the authorization request. 
```

### Scope value "openid" and "claims" parameter

OpenID Connect [@OIDC] specifies the JSON-based `claims` request parameter that can be used to specify the claims a client (acting as OpenID Connect Relying Party) wishes to receive in a fine-grained and privacy preserving way as well as assign those claims to a certain delivery mechanisms, i.e. ID Token or userinfo response. 

> The combination of the scope value `openid` and the additional parameter `claims` can be used beside `authorization_details` in the same way as every other scope value and, potentially, further parameter providing additional data for the respective scope to the authorization process. 

Reviewer: I get what you're saying but this is awkward to add here because it's not defined in OIDC.

```suggestion
The combination of the scope value `openid` and the additional parameter `claims` can be used beside `authorization_details` in the same way as every other non-OIDC scope value. 
```

Alternatively, there could be an authorization data type for OpenID Connect. (#openid) gives an example of how such an authorization data type could look like.

## Relationship to "resource" parameter

The request parameter `resource` [@I-D.ietf-oauth-resource-indicators] indicates to the AS the resource(s) where the client intends to use the access tokens issued based on a certain grant. This mechanism is a way to audience-restrict access tokens and to allow the AS to create resource server specific access tokens. 

> If a client uses `authorization_details` with `locations` elements in an authorization request, it MAY omit the `resource` parameter since the information conveyed in the `authorization_details` object are a superset of what can be conveyed in the `resource` parameter.

Reviewer 2: The `resource` parameter is already optional so this MAY feels out of place here. Maybe combine this with the following paragraph and say something to the effect of oauth-security-topics recommends using `resource` but that recommendation could also be met by use of `locations` in the objects in the `authorization_details`.

Reviewer 2: Or maybe I don't get it...

> Clients requiring audience restricted access tokens are RECOMMENDED to use the `resource` parameter in token requests to allow the AS to narrow down the privileges of the access token to specific permissions for individual operations on specific resource servers (see [@I-D.ietf-oauth-security-topics], Section 3.3). 

Reviewer: This seems to contradict the advice in the previous line.

Author: I try to express that "locations" is more specific than "resource", that's why "resource" is not needed in authorization requests. However, in token requests it makes sense to specify the target the access token shall be used at.

Reviewer 2: I'm confused TBH.

Author: about the text or this conversation?

Author: I tried to improve the text. Please review.

Reviewer 2: I'm honestly having a hard time following all the threads and comments and updates but I do think the changes in 5626d118e29e465c57c7af7f0ce33f80d75dd3d3 improve it.

> If the client used `authorization_details` with `locations` elements in the authorization request, the AS MUST utilize this data to filter the authorization data objects applicable to the respective `resource`. This process will use the `resource` string as prefix to filter the `locations` elements of the authorization details elements.

Reviewer: So `resource` is used as a string-prefix for `locations` values? It's an interesting idea but I'm not convinced I like it.

Reviewer 2: Not sure I'm too crazy about it either. What's the perceived value of the prefixing? It seems to introduce some processing complexity and not-necessary intuitive interdependence between parameters. Also how is is it expected to work, if there are multiple occurrences of `resource`?

> Given the example in (#authz_details), a client could request an access token using a `resource` value of `https://example.com/payments`, which would produce an access token containing the `payment_initation` but not the `account_information` authorization details as shown in the following example:

Reviewer 2: I read this several times before understanding it (and not sure I do now). But maybe clarify that the example in example in (#authz_details) is the authorization request and a you are (presumably?) talking about a subsequent access token request with a refresh token and the resultant access token would contain or reference the JSON below as some claim value. If I even understood that all correctly...

Reviewer 2: Also I'm not a huge fan of having this place for application/API specific authz details but that it might have some of these predefined data elements types and one of those elements has specific behavior relative to a request parameter in the token request.

Author: I tried to improve the text. Please review.

Reviewer 2: I'm honestly having a hard time following all the threads and comments and updates but I do think the changes in 5626d118e29e465c57c7af7f0ce33f80d75dd3d3 improve it.

Author: I would suggest to merge the PR after we have came to a conclusion regarding scope/authz details merge and review the result again.

```JSON
[
   {
      "type": "payment_initiation",
      "actions": [
         "initiate",
         "status",
         "cancel"
      ],
      "locations": [
         "https://example.com/payments"
      ],
      "instructedAmount": {
         "currency": "EUR",
         "amount": "123.50"
      },
      "creditorName": "Merchant123",
      "creditorAccount": {
         "iban": "DE02100100109307118603"
      },
      "remittanceInformationUnstructured": "Ref Number Merchant"
   }
]
```

# Using "authorization_details"

## Authorization Request

The request parameter can be used to specify authorization requirements in all places where the `scope` parameter is used for the same purpose, examples include:  

* Authorization requests as specified in [@!RFC6749], 
* Access token requests as specified in [@!RFC6749], if also used as authorization requests, e.g. in the case of assertion grant types [@!RFC7521],
* Request objects as specified in [@I-D.ietf-oauth-jwsreq], 
* Device Authorization Request as specified in [@!RFC8628],
* Backchannel Authentication Requests as defined in [@OpenID.CIBA].

Parameter encoding is determined by the respective context. 

> In the context of an authorization request according to [@!RFC6749], the parameter is encoded using the `application/x-www-form-urlencoded` format as shown in the following example:

Reviewer: ```suggestion
In the context of an authorization request according to [@!RFC6749], the parameter is encoded using the `application/x-www-form-urlencoded` format of the JSON string as shown in the following example:
```

Reviewer 2: +1

```
GET /authorize?response_type=code
   &client_id=s6BhdRkqt3
   &state=af0ifjsldkj
   &redirect_uri=https%3A%2F%2Fclient.example.org%2Fcb
   &code_challenge_method=S256
   &code_challenge=K2-ltc83acc4h0c9w6ESC_rEMTJ3bww-uCHaoeK1t8U
   &authorization_details=%5B%7B%22type%22%3A%22account%5Finformati
   on%22%2C%22actions%22%3A%5B%22list%5Faccounts%22%2C%22read%5Fbal
   ances%22%2C%22read%5Ftransactions%22%5D%2C%22locations%22%3A%5B%
   22https%3A%2F%2Fexample%2Ecom%2Faccounts%22%5D%7D%5D HTTP/1.1
Host: server.example.com
``` 

Implementors MUST ensure to protect personal identifiable information
in transit. One way is to utilize encrypted request objects as defined
in [@I-D.ietf-oauth-jwsreq]. In the context of a request object, 

> `authorization_details` is added as another top level JSON element.

Reviewer: ```suggestion
`authorization_details` is added as a top level JSON element, just as any other authorization request parameter.
```

   "authorization_details": [
      {
         "type": "https://www.someorg.com/payment_initiation",
         "actions": [
            "initiate",
            "status",
            "cancel"
         ],
         "locations": [
            "https://example.com/payments"
         ],
         "instructedAmount": {
            "currency": "EUR",
            "amount": "123.50"
         },
         "creditorName": "Merchant123",
         "creditorAccount": {
            "iban": "DE02100100109307118603"
         },
         "remittanceInformationUnstructured": "Ref Number Merchant"
      }
   ]
}
```

### Token Content {#token_content}

In order to enable the RS to enforce the authorization details as approved in the authorization process, the AS MUST make this data available to the RS. 

If the access token is a JWT [@!RFC7519], the AS is RECOMMENDED to add the `authorization_details` object, filtered to the specific audience, as top-level claim. 

The AS will typically also add further claims to the JWT the RS requires for request processing, e.g., user id, roles, and transaction specific data. What claims the particular RS requires is defined by the RS-specific policy with the AS.

> The following shows the contents of an example JWT for the payment initation example above: 

Reviewer: ```suggestion
The following shows the contents of an example JWT payload for the payment initation example above: 
```

```JSON
{
   "iss": "https://as.example.com",
   "sub": "24400320",

>    "aud": "s6BhdRkqt3",

Reviewer 2: the client id used throughout the examples elsewhere in the document is the audience here?

Author: fixed

   "exp": 1311281970,
   "acr": "psd2_sca",
   "txn": "8b4729cc-32e4-4370-8cf0-5796154d1296",
   "authorization_details": [
      {
         "type": "https://www.someorg.com/payment_initiation",
         "actions": [
            "initiate",
            "status",
            "cancel"
         ],
         "locations": [
            "https://example.com/payments"
         ],
         "instructedAmount": {
            "currency": "EUR",
            "amount": "123.50"
         },
         "creditorName": "Merchant123",
         "creditorAccount": {
            "iban": "DE02100100109307118603"
         },
         "remittanceInformationUnstructured": "Ref Number Merchant"
      }
   ]
}
```

In this case, the AS added the following example claims:

* `sub`: conveys the user on which behalf the client is asking for payment initation
* `txn`: transaction id used to trace the transaction across the services of provider `example.com`

> * `debtorAccount`: API-specific element containing the debtor account as selected by the user during the authorization process. The field `user_role` conveys the role the user has with respect to this particuar account. In this case, she is the owner. This data is used for access control at the payment API (the RS).

Reviewer 2: I do not see debtorAccount in the above example?

Author: because the debtorAccount was selected by the user in the course of the authorization process

## Token Introspection Request

In case of opaque access tokens, the data provided to a certain RS is determined using the RS's identifier with the AS (see [@I-D.ietf-oauth-jwt-introspection-response], section 3). 

## Token Introspection Response

> The token endpoint response provides the RS with the authorization details applicable to it as a top-level JSON element along with the claims the RS requires for request processing. 

Reviewer: ```suggestion
The token introspection response provides the RS with the authorization details applicable to it as a top-level JSON element along with the claims the RS requires for request processing. 
```

Here is an example for the payment initation example RS:

```json
{
   "active": true,
   "sub": "24400320",

>    "aud": "s6BhdRkqt3",

Reviewer 2: the client id used throughout the examples elsewhere in the document is the audience here?

Author: fixed

   "exp": 1311281970,
   "acr": "psd2_sca",
   "txn": "8b4729cc-32e4-4370-8cf0-5796154d1296",
   "authorization_details": [
      {
         "type": "https://www.someorg.com/payment_initiation",
         "actions": [
            "initiate",
            "status",
            "cancel"
         ],
         "locations": [
            "https://example.com/payments"
         ],
         "instructedAmount": {
            "currency": "EUR",
            "amount": "123.50"
         },
         "creditorName": "Merchant123",
         "creditorAccount": {
            "iban": "DE02100100109307118603"
         },
         "remittanceInformationUnstructured": "Ref Number Merchant"
      }
   ],
   "debtorAccount": {
      "iban": "DE40100100103307118608",
      "user_role": "owner"
   }
}
```

> The differences to the example given in (#token_content) lay in the difference between JWT-encoded access tokens and token introspection responses.

Reviewer: This line is not necessary, suggest we remove it.

```suggestion
```

Author: @b---c what do you think?

Reviewer 2: Seems like it's already implied by the section titles.

The following example shows how an implementation could utilize the namespace `https://scheme.example.org/` to ensure collision resistant element names.

```JSON
{
   "type": "https://scheme.example.org/files",
   "locations": [
      "https://example.com/files"
   ],
   "permissions": [
      {
         "path": "/myfiles/A",
         "access": [
            "read"
         ]
      },
      {
         "path": "/myfiles/A/X",
         "access": [
            "read",
            "write"
         ]
      }
   ]
}
```
## Relationship to "scope" parameter {#scope}

`authorization_details` and `scope` can be used in the same authorization request for carrying independent authorization requirements. 

The AS MUST consider both sets of requirements in combination with each other for the given authorization request. The details of how the AS combines these parameters are specific to the APIs being protected and outside the scope of this specification.

It is RECOMMENDED that a given API use only one form of requirement specification. 

> When gathering user consent, the AS MUST present the merged set of requirements represented by the authorization request. 

Reviewer 2: ```suggestion
When gathering user consent, the AS MUST present the complete set of requirements represented by the authorization request. 
```

   ]
}
```
## Relationship to "scope" parameter

`authorization_details` can be used besides the `scope` request parameter or as a replacement. If both parameters are used in the same request, the authorization a client asks for is determined by the combination of both parameter values.

## Relationship to "resource" parameter

The request parameter `resource` [@I-D.ietf-oauth-resource-indicators] indicates to the AS the resource(s) where the client intends to use the access tokens issued based on a certain grant. This mechanism is a way to audience-restrict access tokens and to allow the AS to create resource server specific access tokens. 

> If a client uses `authorization_details` with `locations` elements in an authorization request, it MAY omit the `resource` parameter since the information conveyed in the `authorization_details` object are a super-set of what can be conveyed in the `resource` parameter.

Author: significantly reworded

Clients requiring audience restricted access tokens are RECOMMENDED to use the `resource` parameter in token requests to allow the AS to narrow down the privileges of the access token to specific permissions for individual operations on specific resource servers (see [@I-D.ietf-oauth-security-topics], section-3.3). 

> If the client used `authorization_details` with `locations` elements in the authorization request, the AS MUST utilize this data to filter the authorization data objects applicable to the respective `resource`. This process will use the `resource` string as prefix to filter the `location` elements of the authorization details elements.

Author: I think the client should always include a single resource value in the token request.
@b---c I re-read draft-ietf-oauth-resource-indicators-08 but I'm not sure whether it allows multiple resource values to turn up in the token request.

The following example shows how a client could ask for OpenID Connect claims and access to account information in the same authorization request. 

```json
[
    {
        "type": "openid_claims",
        "userinfo": {
            "given_name": {
                "essential": true
            },
            "nickname": null,
            "email": {
                "essential": true
            },
            "email_verified": {
                "essential": true
            },
            "picture": null,
            "http://example.info/claims/groups": null
        },
        "id_token": {
            "auth_time": {
                "essential": true
            },
            "acr": {
                "values": [
                    "urn:mace:incommon:iap:silver"
                ]
            }
        }
    },
    {
        "type": "account_information",
        "actions": [
            "list_accounts",
            "read_balances",
            "read_transactions"
        ],
        "locations": [
            "https://example.com/accounts"
        ]
    }

> ]

Reviewer: That's my reading of it -- at least, not without also adding the `openid` scope to the request.

    }
```

### Token Content {#token_content}

In order to enable the RS to enforce the authorization details as approved in the authorization process, the AS MUST make this data available to the RS. 

In case of a JWT [@!RFC7519], the AS is RECOMMENDED to add the `authorization_details` object, filtered to the specific audience, as top-level claim. 

The AS will typically also add further claims to the JWT the RS requires for request processing, e.g. user id, roles and transaction specific data. What claims the particular RS requires is defined by the RS-specific policy with the AS.

> The following shows an example JWT for the payment initation example above: 

Author: sure it does ;-)

> The following shows an example JWT for the payment initation example above: 

Author: Most people even don't understand the differentiation between JWT content the actual JWT.

> The following shows an example JWT for the payment initation example above: 

Author: changed it as you suggested

        <front>
          <title abbrev="CIBA">OpenID Connect Client Initiated Backchannel Authentication Flow - Core 1.0</title>
          <author fullname="Gonzalo Fernandez Rodriguez" initials="G." surname="Fernandez">
            <organization abbrev="Telefonica">Telefonica I+D</organization>
          </author>
          <author fullname="Florian Walter" initials="F." surname="Walter">
            <organization abbrev="">Deutsche Telekom AG</organization>
          </author>
          <author fullname="Axel Nennker" initials="A." surname="Nennker">
            <organization abbrev="">Deutsche Telekom AG</organization>
          </author>
          <author fullname="Dave Tonge" initials="D." surname="Tonge">
            <organization abbrev="Moneyhub">Moneyhub</organization>
          </author>
          <author fullname="Brian Campbell" initials="B." surname="Campbell">
            <organization abbrev="Ping Identity">Ping Identity</organization>
          </author>
          <date day="16" month="January" year="2019"/>
        </front>
      </reference>

{backmatter}

# Additional Examples

## OpenID Connect {#openid}

> This examples tries to encapsulate all details specific to the OpenID Connect part of an authorization process into an authorization JSON object.

Reviewer 2: I fear this will be taken as yet another way to make OIDC requests.

Author: that's what it is - the alternative is to stay with scope forever to turn on openid

Author: I see. If you think it's to sensitiv, we can also drop the example. FAPI could work an a OpenID extension to support authorization_details.

Reviewer 2: I'd be happy with just the change (i.e., change "This examples" to "This hypothetical example") mentioned above.

Author: changed

   }
]
```

The top-level elements have the following meaning:

* `credentialID`: identifier of the certificate to be used for signing
* `documentDigests`: array containing the hash of every document to be signed (`hash` elements). Additionally, the corresponding `label` element identifies the respective document to the user, e.g. to be used in user consent.
* `hashAlgorithm`: algomrithm that was used to calculate the hash values. 

The AS is supposed to ask the user for consent for the creation of signatues for the documents listed in the structure. The client uses the access token issued as result of the process to call the sign doc endpoint at the respective signing service to actually create the signature. This access token is bound to the client, the user id and the hashes (and signature algorithm) as consented by the user.

## Access to Tax Data {#tax}

This example is inspired by an API allowing third parties to access citizen's tax declarations and statements, for example to determine their credit worthiness.

```json
[
    {
        "type": "tax_data",
        "locations": [
            "https://taxservice.govehub.no"
        ],

>         "actions":"read_tax_statement",

Reviewer: ```suggestion
        "actions":["read_tax_statement"],
```

        <front>
          <title abbrev="CIBA">OpenID Connect Client Initiated Backchannel Authentication Flow - Core 1.0</title>
          <author fullname="Gonzalo Fernandez Rodriguez" initials="G." surname="Fernandez">
            <organization abbrev="Telefonica">Telefonica I+D</organization>
          </author>
          <author fullname="Florian Walter" initials="F." surname="Walter">
            <organization abbrev="">Deutsche Telekom AG</organization>
          </author>
          <author fullname="Axel Nennker" initials="A." surname="Nennker">
            <organization abbrev="">Deutsche Telekom AG</organization>
          </author>
          <author fullname="Dave Tonge" initials="D." surname="Tonge">
            <organization abbrev="Moneyhub">Moneyhub</organization>
          </author>
          <author fullname="Brian Campbell" initials="B." surname="Campbell">
            <organization abbrev="Ping Identity">Ping Identity</organization>
          </author>
          <date day="16" month="January" year="2019"/>
        </front>
      </reference>

{backmatter}

# Additional Examples

## OpenID Connect {#openid}

> This hypothetical examples tries to encapsulate all details specific to the OpenID Connect part of an authorization process into an authorization JSON object.

Reviewer 2: ```suggestion
These hypothetical examples try to encapsulate all details specific to the OpenID Connect part of an authorization process into an authorization JSON object.
```
